Refactor editor `EditorBottomPanel` to be a `TabContainer`

lodetrick opened the pull request:
See: https://github.com/godotengine/godot-proposals/issues/12430, https://github.com/godotengine/godot-proposals/issues/12402
Depends on #106263, #103478, #107395

Currently the Bottom Panel inherits from PanelContainer. This makes it hard to integrate with systems like the editor docks that need to make exceptions everywhere in their code for the bottom panel. This PR converts the type of `EditorBottomPanel` to a `TabContainer`. This removes some complexity inside the class as it essentially re-implemented the functionality, but more importantly allows future integrations with the dock system to be much more painless.

MewPurPur commented:
Does this change anything for the user, like functionality or UI?

lodetrick commented:
> Does this change anything for the user, like functionality or UI?

Nothing intended in this PR. Any visual changes are likely bugs. This is just scaffolding for future PRs that would be able to integrate the bottom panel more tightly with the dock system

KoBeWi commented:
I looked at the new BottomPanel's structure and it's basically the same... Still has internal PanelContainer and a ScrollContainer with Buttons. The TabContainer's TabBar is hidden and unused.

That looks hacky to me with no clear benefit. What complexity did you remove if you had to replace TabContainer's default behavior?

lodetrick commented:
It was admittedly hacky to get the custom look of the TabBar and keeping it updated. Thats what the `_notify` methods are mainly for; updating the bottom bar to be consistent with how the TabContainer understands the tabs. The underlying logic and data is transferred to the TabContainer.

KoBeWi commented:
> The main complexity that was lost is that the BottomPanel no longer has to manage its children directly

It didn't do anything more than toggling visibility, no?

I think if you want to refactor the bottom panel into TabContainer, it should become actual TabContainer. You can add a TabBar theme variation that makes it look like the current bottom panel's bar. Although that sounds like a change that should be discussed more.

Removing the buttons would also be useful if we eventually decided to improve docking system and allow moving tabs between docks and bottom panel.

lodetrick commented:
Ok. I tried to create a theme that mimicked the look of the bottom panel and it does look good, the one constraint that I ran into is that the `TabBar` does not have a selected_hovered stylebox, or at least does not show the `hovered` stylebox when hovering over the selected tab. This would probably have to be introduced or changed if we are looking for visual parity with the current master.

lodetrick commented:
Alright, so I've gotten rid of the extraneous code and added a new theme override that mimics master's look pretty well, only having to compromise by having selected panels be "pushed down" because pressed panels can't be hovered. The bottom panel now looks like this:

<img width="1066" alt="New look using theme" src="https://github.com/user-attachments/assets/f32362a0-80e0-486d-ab70-d3675a4c0ad8" />

This refactor has drastically reduced the amount of code in the `EditorBottomPanel` class. That being said, I'm having trouble considering compatibility. The `add_item` method originally returned a button that plugins hid and toggled to their will. I ended up creating a dummy button that has signals connected to functionality that mirrors how it was used in the past.

I would love to deprecate the functions that have become wrappers for `remove_child` or `move_child`, but I'll let that be a discussion for a future PR.

kitbdev commented:
- Also needs #103478 to keep the same behavior
 - https://github.com/godotengine/godot-proposals/issues/2250 would be useful for the buttons
 - Since the buttons could previously have different styles something like https://github.com/godotengine/godot-proposals/issues/9161 is needed too

More options for individual tabs is important to keep it the same, since it is already used for the Output and Debugger to show more information:

![image](https://github.com/user-attachments/assets/7b5a02a4-b2f7-419b-ba00-52e1766a6588)


The tab bar and tab container bug fixes can be moved to a new PR, it would be easier to get them merged and keeps this one more focused.

KoBeWi commented:
Aside from what was said above, the style needs tweaking:
![image](https://github.com/user-attachments/assets/ff7a7900-e5f3-49f6-a5bd-760db7816cf1)
It would be better if the tabs were higher and the text was centered. Does not look good right now.

lodetrick commented:
Right now TabContainer theming is really limited when it comes to vertical margins. If #106421 was fixed then this would be able to be done a lot easier.

KoBeWi commented:
I'm not sure about the new pressed style:
![image](https://github.com/user-attachments/assets/d8dd8322-68e7-4159-9fde-dbf2122cea1b)
But other than that the changed bottom panel looks the same, which is good.

KoBeWi commented on editor/debugger/editor_debugger_node.cpp:
```diff
@@ -435,26 +435,34 @@ void EditorDebuggerNode::_update_errors() {
+		last_error_count = error_count;
+		last_warning_count = warning_count;
+
+		TabContainer *parent = static_cast<TabContainer *>(get_parent());
```
```suggestion
		TabContainer *parent = Object::cast_to<TabContainer>(get_parent());
```

KoBeWi commented:
Crash when using Find in Files:
```
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.5.dev.custom_build (1b37dacc1842779fb0d03a5b09026f59c13744fc)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
[0] Node::is_accessible_from_caller_thread (C:\godot_source\scene\main\node.h:657)
[1] Node::is_accessible_from_caller_thread (C:\godot_source\scene\main\node.h:657)
[2] Node::get_child_count (C:\godot_source\scene\main\node.cpp:1916)
[3] ScriptEditor::_start_find_in_files (C:\godot_source\editor\plugins\script_editor_plugin.cpp:4081)
[4] call_with_variant_args_helper<ScriptEditor,bool,0> (C:\godot_source\core\variant\binder_common.h:228)
[5] call_with_variant_args<ScriptEditor,bool> (C:\godot_source\core\variant\binder_common.h:338)
[6] CallableCustomMethodPointer<ScriptEditor,void,bool>::call (C:\godot_source\core\object\callable_method_pointer.h:107)
[7] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[8] CallableCustomBind::call (C:\godot_source\core\variant\callable_bind.cpp:151)
[9] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[10] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1288)
[11] Node::emit_signalp (C:\godot_source\scene\main\node.cpp:4287)
[12] Object::emit_signal<> (C:\godot_source\core\object\object.h:940)
[13] FindInFilesDialog::custom_action (C:\godot_source\editor\find_in_files.cpp:606)
[14] FindInFilesDialog::_on_search_text_submitted (C:\godot_source\editor\find_in_files.cpp:625)
[15] call_with_variant_args_helper<FindInFilesDialog,String const &,0> (C:\godot_source\core\variant\binder_common.h:223)
[16] call_with_variant_args<FindInFilesDialog,String const &> (C:\godot_source\core\variant\binder_common.h:338)
[17] CallableCustomMethodPointer<FindInFilesDialog,void,String const &>::call (C:\godot_source\core\object\callable_method_pointer.h:107)
[18] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[19] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1288)
[20] Node::emit_signalp (C:\godot_source\scene\main\node.cpp:4287)
[21] Object::emit_signal<String> (C:\godot_source\core\object\object.h:940)
[22] LineEdit::gui_input (C:\godot_source\scene\gui\line_edit.cpp:875)
[23] Control::_call_gui_input (C:\godot_source\scene\gui\control.cpp:1866)
[24] Viewport::_gui_input_event (C:\godot_source\scene\main\viewport.cpp:2274)
[25] Viewport::push_input (C:\godot_source\scene\main\viewport.cpp:3427)
[26] Window::_window_input (C:\godot_source\scene\main\window.cpp:1839)
[27] call_with_variant_args_helper<Window,Ref<InputEvent> const &,0> (C:\godot_source\core\variant\binder_common.h:223)
[28] call_with_variant_args<Window,Ref<InputEvent> const &> (C:\godot_source\core\variant\binder_common.h:338)
[29] CallableCustomMethodPointer<Window,void,Ref<InputEvent> const &>::call (C:\godot_source\core\object\callable_method_pointer.h:107)
[30] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[31] Callable::call<Ref<InputEvent> > (C:\godot_source\core\variant\variant.h:953)
[32] DisplayServerWindows::_dispatch_input_event (C:\godot_source\platform\windows\display_server_windows.cpp:4387)
[33] DisplayServerWindows::_dispatch_input_events (C:\godot_source\platform\windows\display_server_windows.cpp:4358)
[34] Input::_parse_input_event_impl (C:\godot_source\core\input\input.cpp:903)
[35] Input::flush_buffered_events (C:\godot_source\core\input\input.cpp:1184)
[36] DisplayServerWindows::process_events (C:\godot_source\platform\windows\display_server_windows.cpp:3777)
[37] OS_Windows::run (C:\godot_source\platform\windows\os_windows.cpp:2253)
[38] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:97)
[39] _main (C:\godot_source\platform\windows\godot_windows.cpp:122)
[40] main (C:\godot_source\platform\windows\godot_windows.cpp:136)
[41] WinMain (C:\godot_source\platform\windows\godot_windows.cpp:150)
[42] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[43] <couldn't map PC to fn name>
-- END OF C++ BACKTRACE --
```
The editor assumes the button has a parent.

lodetrick commented:
The crash should be fixed, and I looked through the rest of the usages of the button and they were fine. Regarding the Debugger Node margins, I fixed that in master in a separate PR, but noticed that with this refactor the fix is a lot simpler and removes lines of code, so I went with that.

Regarding the theming, the TabBar class doesn't have a theming option for `pressed_hover`, so the pressed look needs to be different from unpressed in order to not have the hover look weird when the mouse is moving over the pressed tab. I'd be open to a separate color, maybe something in between the background and what there is now.

Mickeon commented on editor/editor_dock_manager.h:
```diff
@@ -75,6 +76,7 @@ class EditorDockManager : public Object {
 
 private:
 	friend class DockContextPopup;
+	friend class EditorBottomPanel; // Temporary until DOCK_SLOT_BOTTOM registered. Used to connect signals.
```
What would fit this comment best, to keep track of it better? TODO? HACK?

Mickeon commented on editor/editor_log.cpp:
```diff
@@ -254,8 +255,13 @@ void EditorLog::add_message(const String &p_msg, MessageType p_type) {
-void EditorLog::set_tool_button(Button *p_tool_button) {
-	tool_button = p_tool_button;
+void EditorLog::_set_dock_tab_icon(Ref<Texture2D> p_icon) {
+	// This is the sole reason to import TabContainer here.
```
You mean...?
```suggestion
	// This is the sole reason to include "tab_container.h" here.
```
Or similar

Mickeon commented on editor/gui/editor_bottom_panel.cpp:
```diff
@@ -32,310 +32,191 @@
+	}
+
+	// Still return a dummy button for compatibility reasons.
+	Button *tb = memnew(Button);
```
Oof, hold on. If not handled properly this button will now either leak memory, or cause errors when attempting to remove it from the tree. Although I suppose that the latter was never really recommended to do as an user. But what about the memory leak? Are the legacy buttons freed appropriately at every possible occasion they should?

lodetrick replied:
Regarding memory leaks, I made sure to memdelete or queue_free all of the buttons in the BottomPanel's destructor and when removing docks, so it shouldn't be an issue.

lodetrick commented:
Ok, I tried to compile a list of feasible editor colors. Personally the top three seem to be the most reasonable but I'm curious what you all think

The color used by selected buttons (current)
<img width="456" alt="BottomPanel(button_selected)" src="https://github.com/user-attachments/assets/29058008-1a92-45d4-9ef6-95fdc79e6261" />

dark_color_1
<img width="394" alt="BottomPanel(dark_color_1)" src="https://github.com/user-attachments/assets/5d207b2a-d86e-4839-86ef-03a508558b00" />

dark_color_2
<img width="396" alt="BottomPanel(dark_color_2)" src="https://github.com/user-attachments/assets/55ae375b-22b6-46e9-a836-1d1af78d5dbe" />

dark_color_3
<img width="401" alt="BottomPanel(dark_color_3)" src="https://github.com/user-attachments/assets/5f8ee0e5-b6e7-494c-96bf-eae312fa2f10" />

highlight_color
<img width="400" alt="BottomPanel(highlight_color)" src="https://github.com/user-attachments/assets/e6f1aab1-aa38-4e8e-8f81-eacf5ecec8d0" />

KoBeWi commented on editor/editor_dock_manager.h:
```diff
@@ -30,6 +30,7 @@
 
 #pragma once
 
+#include "gui/editor_bottom_panel.h"
```
Should be in cpp.

KoBeWi commented on editor/plugins/shader_editor_plugin.cpp:
```diff
@@ -938,7 +938,7 @@ ShaderEditorPlugin::ShaderEditorPlugin() {
 	empty.instantiate();
 	shader_tabs->add_theme_style_override(SceneStringName(panel), empty);
 
-	button = EditorNode::get_bottom_panel()->add_item(TTR("Shader Editor"), window_wrapper, ED_SHORTCUT_AND_COMMAND("bottom_panels/toggle_shader_editor_bottom_panel", TTRC("Toggle Shader Editor Bottom Panel"), KeyModifierMask::ALT | Key::S));
```
I like how you deleted the button and it required no other changes XD

KoBeWi commented on editor/gui/editor_bottom_panel.cpp:
```diff
@@ -32,310 +32,191 @@
-	if (p_at_front) {
-		button_hbox->move_child(tb, 0);
+void EditorBottomPanel::make_item_visible(Control *p_item, bool p_visible, bool p_ignore_lock) {
+	// Don't allow changing tabs involuntarily when tabs are locked
```
```suggestion
	// Don't allow changing tabs involuntarily when tabs are locked.
```

KoBeWi commented on editor/gui/editor_bottom_panel.cpp:
```diff
@@ -32,310 +32,191 @@
+	bottom_docks.remove_at(item_idx);
+	dock_shortcuts.remove_at(item_idx);
+
+	legacy_buttons[item_idx]->queue_free();
```
This can crash if item is not found.

lodetrick commented on editor/gui/editor_bottom_panel.cpp:
```diff
@@ -32,310 +32,191 @@
+		// Hide panel when not showing anything.
+		add_theme_style_override(SceneStringName(panel), get_theme_stylebox("BottomPanelHidden", EditorStringName(EditorStyles)));
+	} else {
+		if (EditorDebuggerNode::get_singleton() == get_current_tab_control()) {
```
I could have the debugger node set its margins, but it also seems really hacky (and I would need to set the corner radius to match the BottomPanel):
```
add_theme_constant_override("margin_left", -EditorNode::get_singleton()->get_editor_theme()->get_stylebox(SNAME("BottomPanel"), EditorStringName(EditorStyles))->get_margin(SIDE_LEFT));
add_theme_constant_override("margin_right", -EditorNode::get_singleton()->get_editor_theme()->get_stylebox(SNAME("BottomPanel"), EditorStringName(EditorStyles))->get_margin(SIDE_RIGHT));
add_theme_constant_override("margin_bottom", -EditorNode::get_singleton()->get_editor_theme()->get_stylebox(SNAME("BottomPanel"), EditorStringName(EditorStyles))->get_margin(SIDE_BOTTOM));
add_theme_constant_override("margin_top", -EditorNode::get_singleton()->get_editor_theme()->get_stylebox(SNAME("BottomPanel"), EditorStringName(EditorStyles))->get_margin(SIDE_TOP));
```

KoBeWi replied:
Well the benefit is that the debugger manages itself this way, so you don't have a special case for a dock.
Though some more universal solution would be nice. What if a user runs into the same problem with their plugin?

lodetrick replied:
Maybe this could be included as a toggle in the EditorDock? Since it extends MarginContainer like the DebuggerNode currently does

KoBeWi replied:
Why does debugger have margins? If they are the same as in bottom panel, maybe the should be removed?
Or, you could add a special case for MarginContainer and cancel out its margins maybe. Still hacky, but at least not towards single dock.

lodetrick replied:
For reference, this is what the debugger would look like if there was no exception made for the debugger node.

<img width="885" alt="Screenshot 2025-06-05 at 3 00 22 PM" src="https://github.com/user-attachments/assets/ff1efe1e-f107-4aa9-9562-62d4fdaf65f2" />

The BottomPanel applies margins to the TabContainer, which looks bad (and pushes the debugger node up). Ideally, the BottomPanel should not have any margin and let the EditorDock class handle that, but for the moment I think that the current hack is the least amount of code to address this without the EditorDock class.

lodetrick commented:
> I don't know how I managed it, but Ctrl+J is altering between two tabs:
> 
>  dzGMy0mf5p.mp4 
> It should toggle one tab on and off.

This is not an issue, because by default Ctrl+J is bound to `Toggle Last Opened Bottom Panel`, so this is intended behavior

KoBeWi commented:
Yeah but it's different behavior than on master.

lodetrick commented:
I'll fix that then. There's a similar question for `EditorBottomPanel::make_item_visible`. If we are trying to hide a dock that is shown, is it preferable to hide the BottomPanel or swap to the previous tab?

lodetrick commented on editor/gui/editor_bottom_panel.cpp:
```diff
@@ -32,310 +32,191 @@
+void EditorBottomPanel::shortcut_input(const Ref<InputEvent> &p_event) {
+	ERR_FAIL_COND(p_event.is_null());
+
+	if (!p_event->is_pressed() && !p_event->is_echo()) {
```
Made it an early return

KoBeWi reviewed (approved):
Seems fine now.

Note that this depends on #106263, #103478 and #107395

lodetrick committed:
Refactor Bottom Panel to be a TabContainer

AThousandShips commented on editor/gui/editor_bottom_panel.h:
```diff
@@ -30,51 +30,41 @@
+#include "scene/gui/tab_container.h"
 
 class Button;
+class Control;
```
```suggestion
```

lodetrick commented on scene/gui/tab_bar.h:
```diff
@@ -222,6 +233,10 @@ class TabBar : public Control {
 	int get_tab_icon_max_width(int p_tab) const;
 
+	void set_font_color_override_all(int p_tab, Color p_color);
+	void set_font_color_override(int p_tab, DrawMode p_draw_mode, Color p_color);
```
Should I put this in #106263 or here?

KoBeWi replied:
In #106263.

lodetrick commented:
Pulled from the updated #106263.

I fixed the sizing issues by setting a minimum size to something that I thought was reasonable. I also realized that there was a hack remaining (overriding `_update_margins()`), so I removed that by replacing it with a custom margin.

With regards to events on mouse pressed vs mouse released, I see what you mean about it being different than before, but that's how `TabBar` is coded so I'm not too sure how possible it would be to change, it would probably be a separate PR at least, it seems like a larger change (though I am curious why it is that way).

arkology commented:
> I fixed the sizing issues by setting a minimum size to something that I thought was reasonable.

I would prefer some more adaptive solution, but it's ok for now.

Found another bug – panel layout breaks when using Arabic (right-to-left language).

KoBeWi commented:
Needs final (?) rebase. Also shouldn't the commits be squashed?

lodetrick commented:
I ran into this same issue when writing #106426, and I solved it with this snippet:

```c++
void EditorBottomPanel::_theme_changed() {
	Ref<StyleBox> bottom_tabbar_style = get_theme_stylebox("tabbar_background", "BottomPanel")->duplicate();
	bottom_tabbar_style->set_content_margin(SIDE_RIGHT, bottom_hbox->get_minimum_size().x + bottom_tabbar_style->get_content_margin(SIDE_LEFT));
	add_theme_style_override("tabbar_background", bottom_tabbar_style);
```

It's been a while since I wrote it but essentially I modified the margins of `tabbar_background` to create the space necessary for the icon to not clip with anything. Another solution would be to just make it a button like the bell and version number instead of the built in tabcontainer use. Would you want me to include this in this PR?

lodetrick commented:
It works by adding the button hbox as a child of the `tab_bar` within the `tabcontainer` so it doesn't get counted as one of the tabs, and then setting the margins to right, expand right. Speaking of that I think it _might_ be possible to get this to work simply if you modify around line 208 in this PR of `editor_bottom_panel.cpp` so that it has an offset equal to the size of the button:
```c++
bottom_hbox->set_anchors_and_offsets_preset(Control::PRESET_RIGHT_WIDE);
```

KoBeWi commented:
> It works by adding the button hbox as a child of the tab_bar within the tabcontainer so it doesn't get counted as one of the tabs, and then setting the margins to right, expand right.

Yeah, but why does TabBar/TabContainer respect that margin? I checked with EditorDebugger, the TabContainer encompasses the buttons
<img width="299" height="154" alt="image" src="https://github.com/user-attachments/assets/afec484a-e37c-4c48-9a1f-7f26865a3c7d" />
but the TabBar is perfectly fit to them:
<img width="461" height="71" alt="image" src="https://github.com/user-attachments/assets/ca27a48d-52dd-48da-9a11-74932c3be500" />
TabContainer will normally stretch the TabBar to full width.

btw, I actually found a bug 💀 The version goes out of bounds if editor is launched with closed bottom panel:
<img width="164" height="80" alt="image" src="https://github.com/user-attachments/assets/f89b9651-de0d-40bb-9e10-dbf0ea4c9c17" />

arkology commented:
> Does that crash occur in this PR? I thought the 3-dot button was disabled in this one.

Oops, sorry 😅 I downloaded PR artifacts from this PR and https://github.com/godotengine/godot/pull/108647 and looks like I mixed up everything. Silly me.

kitbdev commented on editor/debugger/editor_debugger_node.cpp:
```diff
@@ -441,26 +441,35 @@ void EditorDebuggerNode::_update_errors() {
+		last_warning_count = warning_count;
+
+		// TODO: Replace logic when EditorDock class is merged to be more flexible.
+		TabContainer *parent = Object::cast_to<TabContainer>(get_parent());
```
The EditorDock class is already merged.

KoBeWi commented:
The version button bug I mentioned in https://github.com/godotengine/godot/pull/106164#issuecomment-3439078546 could be fixed though.

Repiteo commented:
Thanks!
